File: 14-TypeScript/Grey Multi-Tenant/packages/grey-kotlin/src/main/kotlin/com/grey/sdk/grpc/GreyChannel.kt
// =============================================================================
// Grey Multi-Tenant SDK - Kotlin
// grpc/GreyChannel.kt
//
// Shared gRPC channel management.
// =============================================================================

package com.grey.sdk.grpc

import com.grey.sdk.config.GreyOptions
import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import io.grpc.Metadata
import io.grpc.stub.MetadataUtils
import java.util.concurrent.TimeUnit

/**
 * Manages the gRPC channel for Grey SDK.
 */
class GreyChannel(private val options: GreyOptions) {

    private var channel: ManagedChannel? = null

    /**
     * Get or create the gRPC channel.
     */
    @Synchronized
    fun getChannel(): ManagedChannel {
        if (channel == null || channel!!.isShutdown) {
            channel = createChannel()
        }
        return channel!!
    }

    /**
     * Create a new gRPC channel with current options.
     */
    private fun createChannel(): ManagedChannel {
        val builder = ManagedChannelBuilder
            .forAddress(options.host, options.port)

        if (options.useTls) {
            builder.useTransportSecurity()
        } else {
            builder.usePlaintext()
        }

        return builder.build()
    }

    /**
     * Create metadata with authorization header.
     */
    fun createAuthMetadata(accessToken: String? = options.accessToken): Metadata {
        val metadata = Metadata()
        accessToken?.let {
            metadata.put(
                Metadata.Key.of("Authorization", Metadata.ASCII_STRING_MARSHALLER),
                "Bearer $it"
            )
        }
        return metadata
    }

    /**
     * Shutdown the channel.
     */
    @Synchronized
    fun shutdown() {
        channel?.shutdown()
        try {
            channel?.awaitTermination(5, TimeUnit.SECONDS)
        } catch (e: InterruptedException) {
            channel?.shutdownNow()
        }
        channel = null
    }

    /**
     * Check if the channel is active.
     */
    val isActive: Boolean
        get() = channel != null && !channel!!.isShutdown && !channel!!.isTerminated
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-swift/Sources/GreySDK/GRPC/GreyChannel.swift
// =============================================================================
// Grey Multi-Tenant SDK - Swift
// GRPC/GreyChannel.swift
//
// Shared gRPC channel management.
// =============================================================================

import Foundation
import GRPC
import NIO
import NIOSSL
import NIOTransportServices

/// Manages the gRPC channel for Grey SDK.
public final class GreyChannel: Sendable {
    
    /// The event loop group for async operations.
    private let group: EventLoopGroup
    
    /// The gRPC channel.
    private let channel: GRPCChannel
    
    /// Configuration options.
    private let options: GreyOptions
    
    // MARK: - Initialization
    
    /// Create a new GreyChannel.
    /// - Parameter options: Configuration options
    /// - Throws: If channel creation fails
    public init(options: GreyOptions) throws {
        self.options = options
        
        // Create event loop group
        #if canImport(Network)
        self.group = NIOTSEventLoopGroup()
        #else
        self.group = MultiThreadedEventLoopGroup(numberOfThreads: 1)
        #endif
        
        // Build channel
        let builder: ClientConnection.Builder
        
        if options.useTLS {
            builder = ClientConnection.usingTLSBackedByNIOSSL(on: group)
        } else {
            builder = ClientConnection.insecure(group: group)
        }
        
        self.channel = builder
            .withConnectionTimeout(minimum: .seconds(Int64(options.timeoutSeconds)))
            .connect(host: options.host, port: options.port)
    }
    
    // MARK: - Channel Access
    
    /// Get the underlying gRPC channel.
    public func getChannel() -> GRPCChannel {
        channel
    }
    
    /// Create call options with optional authorization.
    /// - Parameter accessToken: Optional access token
    /// - Returns: CallOptions with timeout and optional auth header
    public func callOptions(accessToken: String? = nil) -> CallOptions {
        var options = CallOptions(
            timeLimit: .timeout(.seconds(Int64(self.options.timeoutSeconds)))
        )
        
        if let token = accessToken ?? self.options.accessToken {
            options.customMetadata.add(name: "authorization", value: "Bearer \(token)")
        }
        
        return options
    }
    
    // MARK: - Lifecycle
    
    /// Shutdown the channel and release resources.
    public func shutdown() async throws {
        try await channel.close().get()
        try await group.shutdownGracefully()
    }
    
    /// Check if the channel is active.
    public var isActive: Bool {
        // Channel connectivity check
        true // Simplified - in production, check actual state
    }
}
